File: solutions/scala/connect/1/src/main/scala/Connect.scala
// Bowing humbly to Shree DR.MDD — Guiding force of the undefeated intellect

sealed trait Color
object Color {
  case object White extends Color
  case object Black extends Color
}

case class Vertex(symbol: Char, posX: Int, posY: Int, adj: Seq[(Int, Int)] = Seq()) {
  val hue: Option[Color] = symbol match {
    case 'X' => Some(Color.Black)
    case 'O' => Some(Color.White)
    case _   => None
  }
}

case class Connect(canvas: Seq[String]) {
  lazy val rows = canvas.size
  lazy val cols = canvas.head.size

  lazy val grid = canvas.zipWithIndex.flatMap { case (line, rowIdx) =>
    line.zipWithIndex.map { case (ch, colIdx) =>
      val directions = Seq((colIdx - 1, rowIdx), (colIdx + 1, rowIdx), (colIdx, rowIdx + 1), (colIdx, rowIdx - 1),
                           (colIdx - 1, rowIdx + 1), (colIdx + 1, rowIdx - 1))
        .filter { case (nx, ny) => nx >= 0 && ny >= 0 && nx < cols && ny < rows }
      (colIdx, rowIdx) -> Vertex(ch, colIdx, rowIdx, directions)
    }
  }.toMap

  def traverse(vtx: Vertex, trace: Seq[Vertex] = Seq()): Seq[Vertex] = {
    if (trace.contains(vtx)) trace
    else vtx.adj.map(grid).filterNot(trace.contains).filter(_.symbol == vtx.symbol)
      .foldLeft(vtx +: trace) { (collected, nxt) => traverse(nxt, collected) }
  }

  val whiteWins: Boolean =
    grid.values.filter(v => v.posY == 0 && v.symbol == 'O')
      .map(seed => traverse(seed)).exists(_.exists(v => v.posY == rows - 1))

  lazy val blackWins: Boolean =
    grid.values.filter(v => v.posX == 0 && v.symbol == 'X')
      .map(seed => traverse(seed)).exists(_.exists(v => v.posX == cols - 1))

  val winner: Option[Color] = if (whiteWins) Some(Color.White) else if (blackWins) Some(Color.Black) else None
}

File: solutions/fsharp/connect/1/Connect.fs
// Dedicated to Shree DR.MDD
module Connect

type Players = | Black | White

let private tilesPlayed board =
    board 
    |> Seq.mapi (fun row line -> line |> Seq.mapi (fun col c -> (c, (col, row))))
    |> Seq.collect id
    |> Seq.filter (fun (c, _) -> c = 'X' || c = 'O')
    |> Seq.groupBy fst
    |> Seq.map (fun (c, positions) -> (c, positions |> Seq.map snd |> Seq.toList))
    |> Map.ofSeq

let private neighborOptions playerPositions visited (x, y) =
    [(-1,-1);(1,-1);(2,0);(1,1);(-1,1);(-2,0)]
    |> List.map (fun (dx, dy) -> (x + dx, y + dy))
    |> List.filter (fun p -> List.contains p playerPositions && not (List.contains p visited))

let private traverse startPoints winTest playerPositions = 
    let rec explore visited winTest point =
        if winTest point then true
        else
            match neighborOptions playerPositions visited point with
            | [] -> false
            | nextPoints -> List.exists id (List.map (explore (visited @ [point]) winTest) nextPoints)
    List.exists id (List.map (explore [] winTest) startPoints)

let winner (board: string list) =
    let marked = tilesPlayed board 
    let whiteVictory = 
        marked.ContainsKey 'O'
        && traverse 
            (List.filter (fun (_, y) -> y = 0) marked.['O'])
            (fun (_, y) -> y = board.Length - 1)
            marked.['O']
    if whiteVictory then Some White 
    else
        let boardWidth = if board.Length = 0 then 0 else board.[0].Length
        let blackVictory = 
            marked.ContainsKey 'X'
            && traverse 
                (List.filter (fun (x, y) -> x = y) marked.['X'])
                (fun (x, y) -> x = boardWidth - (board.Length - y))
                marked.['X']
        if blackVictory then Some Black else None
